src/Stratis.Bitcoin/Consensus/ConsensusManager.cs

```diff
@@ -211,9 +211,63 @@ public void PeerDisconnected(int peerId)
         }
 
         /// <inheritdoc />
-        public Task<ChainedHeaderBlock> BlockMined(Block block)
+        public async Task<ChainedHeaderBlock> BlockMinedAsync(Block block)
         {
-            throw new NotImplementedException();
+            this.logger.LogTrace("({0}:{1})", nameof(block), block.GetHash());
+
+            PartialValidationResult partialValidationResult;
+
+            using (await this.reorgLock.LockAsync().ConfigureAwait(false))
+            {
+                ChainedHeader chainedHeader;
+
+                lock (this.peerLock)
+                {
+                    if (block.Header.HashPrevBlock != this.Tip.HashBlock)
+                    {
+                        this.logger.LogError("(-)[BLOCKMINED_INVALID_PREVIOUS_TIP]");
+                        return null;
+                    }
+
+                    chainedHeader = this.chainedHeaderTree.CreateChainedHeaderWithBlock(block);
+                }
+
+                partialValidationResult = await this.partialValidator.ValidateAsync(block, chainedHeader).ConfigureAwait(false);
+                if (partialValidationResult.Succeeded)
+                {
+                    bool fullValidationRequired;
+
+                    lock (this.peerLock)
+                    {
+                        this.chainedHeaderTree.PartialValidationSucceeded(chainedHeader, out fullValidationRequired);
+                    }
+
+                    if (fullValidationRequired)
+                    {
+                        ConnectBlocksResult fullValidationResult = await this.FullyValidateLockedAsync(partialValidationResult.ChainedHeaderBlock).ConfigureAwait(false);
+                        if (!fullValidationResult.Succeeded)
+                        {
+                            this.logger.LogError("Miner produced an invalid block, full validation failed: {0}", fullValidationResult.Error.Message);
+                            this.logger.LogTrace("(-)[FULL_VALIDATION_FALIED]");
+                            throw new ConsensusException(fullValidationResult.Error.Message);
+                        }
+                    }
+                }
+                else
+                {
+                    lock (this.peerLock)
+                    {
+                        this.chainedHeaderTree.PartialOrFullValidationFailed(chainedHeader);
+                    }
+
+                    this.logger.LogError("Miner produced an invalid block, partial validation failed: {0}", partialValidationResult.Error.Message);
+                    this.logger.LogTrace("(-)[PARTIAL_VALIDATION_FALIED]");
+                    throw new ConsensusException(partialValidationResult.Error.Message);
+                }
+            }
+
+            this.logger.LogTrace("(-)");
```

Comment:
Include return value here:

             this.logger.LogTrace("(-):{0}",…);

Reply:
Fixed